### PR TITLE
Add `area` endpoint support

### PR DESCRIPTION
**Pull request by mpiannucci**
* Add `area` endpoint
* Move selection logic to its own module
* add tests for selection

### CODE REVIEW

File: xpublish_edr/select.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    # Find the nearest X and Y coordinates to the point
+    return ds.cf.sel(X=point.x, Y=point.y, method="nearest")
+
+
+def _select_area_regular_xy_grid(ds: xr.Dataset, polygon: Polygon) -> xr.Dataset:
+    """
+    Return a dataset with the area within the given polygon
+    """
+    # For a regular grid, we can create a meshgrid of the X and Y coordinates to create a spatial mask
+    pts = np.meshgrid(ds.cf["X"], ds.cf["Y"])
```
Comment (dcherian) [inline xpublish_edr/select.py:109]: I would clip these to `polygon.bounds` to minimize the perf hit of `np.vectorize`

https://shapely.readthedocs.io/en/stable/reference/shapely.Polygon.html#shapely.Polygon.bounds

Comment (mpiannucci) [reply xpublish_edr/select.py:109]: great idea, the downselection happens later anyways so might as well do it up front

File: xpublish_edr/select.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    # Create a mask of the points within the polygon
+    contains = np.vectorize(lambda p: polygon.contains(Point(p)), signature="(n)->()")
+    mask = contains(np.stack(pts, axis=-1))
```
Comment (dcherian) [inline xpublish_edr/select.py:113]:
```suggestion
    mask = shapely.contains_xy(polygon, pts[0], pts[1])
```
maybe?

I have learned that shapely has all kinds of useful functions that run the inner in C through GEOS, but are quite hard to find. This is potentially one of them: https://shapely.readthedocs.io/en/stable/reference/shapely.contains_xy.html#shapely.contains_xy

File: xpublish_edr/select.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    # Find the x and y indices that have any points within the polygon
+    x_mask = mask.any(axis=0)
+    y_mask = mask.any(axis=1)
+
+    # Create a new DataArray with the mask matching the dataset dimensions
+    dims = ds.cf["Y"].dims + ds.cf["X"].dims
+    da_mask = xr.DataArray(mask, dims=dims)
+
+    # Apply the mask and subset out any data that does not fall within the polygon bounds
+    ds_subset = ds.where(da_mask).cf.isel(X=x_mask, Y=y_mask)
```
Comment (dcherian) [inline xpublish_edr/select.py:124]: Doesn;t this return a bunch of NaNs for say a pentagon?

IIUC what you need is
```
xinds, yinds = np.argwhere(mask)
return ds.cf.isel(X=xr.Variable(xinds, dims="points"), Y=xr.Variable(yinds, dims="points")
```

to trigger vectorized indexing

File: xpublish_edr/select.py (hunk 2 above)
Comment (dcherian) [reply xpublish_edr/select.py:113]: [shapely.readthedocs.io/en/stable/reference/shapely.contains.html#shapely.contains](https://shapely.readthedocs.io/en/stable/reference/shapely.contains.html#shapely.contains):

"following this definition, a geometry does not contain its boundary, but it does contain itself. "

So... if you request a polygon where a vertex is on the grid, what do we get?

File: xpublish_edr/select.py (hunk 3 above)
Comment (mpiannucci) [reply xpublish_edr/select.py:124]: Can xarray support a staggered array for this? The reason for the currenly impl is because i wasnt changing the coordinates, but maybe i should?

Comment (mpiannucci) [reply xpublish_edr/select.py:124]: Oh this is soooo cool. Thanks @dcherian !!!

Comment (dcherian) [reply xpublish_edr/select.py:124]: oops, I meant `np.nonzero` not `np.argwhere`

File: xpublish_edr/select.py (hunk 2 above)
Comment (mpiannucci) [reply xpublish_edr/select.py:113]: It clips. What do we expect though? My intuition is that it should be included if it hits the boundary? My first thought is a tiny buffer?

Comment (dcherian) [reply xpublish_edr/select.py]: does the standard say anything?

Comment (mpiannucci) [reply xpublish_edr/select.py]: nope

**Comment by mpiannucci** [comment]
pre-commit.ci autofix

File: xpublish_edr/select.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    # To minimize performance impact, we first subset the dataset to the bounding box of the polygon
+    polygon = polygon.buffer(buffer)
+    (minx, miny, maxx, maxy) = polygon.bounds
+    ds = ds.cf.sel(X=slice(minx, maxx), Y=slice(maxy, miny))
+
+    # For a regular grid, we can create a meshgrid of the X and Y coordinates to create a spatial mask
+    pts = np.meshgrid(ds.cf["X"], ds.cf["Y"])
+
+    # Create a mask of the points within the polygon
+    mask = shapely.contains_xy(polygon, pts[0], pts[1])
```
Comment (dcherian) [inline xpublish_edr/select.py:128]:
```suggestion
    mask = shapely.intersects_xy(polygon, pts[0], pts[1])
```

https://shapely.readthedocs.io/en/stable/reference/shapely.intersects.html#shapely.intersects
> Intersects implies that overlaps, touches, covers, or within are True.

do you have a test for this edge case?

```python
poly = shapely.Polygon.from_bounds(0, 1, 2, 3)
shapely.intersects_xy(poly, x=[0, 0.5, 1], y=[2, 2.5, 3])
# array([ True,  True,  True])
```

Comment (mpiannucci) [reply xpublish_edr/select.py:128]: yeah in `test_select.py`

**Comment by mpiannucci** [comment]
pre-commit.ci autofix

**Comment by mpiannucci** [comment]
Love when codecoverage goes down from adding more tests lol

File: xpublish_edr/plugin.py [hunk 5]
```diff
@@ -79,65 +93,71 @@ def get_position(
...
-            if query.z:
-                ds = dataset.cf.sel(Z=query.z, method="nearest")
+            logger.debug(
+                f"Dataset filtered by position ({query.geometry.x}, {query.geometry.y}): {ds}",
+            )
 
-            if query.datetime:
-                datetimes = query.datetime.split("/")
+            try:
+                ds = select_query(ds, query, dict(request.query_params))
```
Comment (abkfenris) [inline xpublish_edr/plugin.py:108]:
```suggestion
                ds = select_by_query_params(ds, query, dict(request.query_params))
```

File: xpublish_edr/plugin.py [hunk 6]
```diff
@@ -79,65 +93,71 @@ def get_position(
             Extra selecting/slicing parameters can be provided as extra query parameters
             """
             try:
-                ds = dataset.cf.sel(X=query.point.x, Y=query.point.y, method="nearest")
+                ds = select_postition(dataset, query.geometry)
```
Comment (abkfenris) [inline xpublish_edr/plugin.py:96]:
```suggestion
                ds = select_by_postition(dataset, query.geometry)
```

File: xpublish_edr/plugin.py [hunk 7]
```diff
@@ -79,65 +93,71 @@ def get_position(
...
-                    continue
-                elif len(split_value) == 2:
-                    query_params[key] = slice(split_value[0], split_value[1])
-                    method = None
-                else:
-                    raise HTTPException(404, f"Too many values for selecting {key}")
+            Extra selecting/slicing parameters can be provided as extra query parameters
+            """
+            try:
+                ds = select_area(dataset, query.geometry)
```
Comment (abkfenris) [inline xpublish_edr/plugin.py:143]:
```suggestion
                ds = select_by_area(dataset, query.geometry)
```

**Review by abkfenris — APPROVED** [review]
I think this generally looks good, but `select.py` and it's functions are a little more complicated to follow than it feels like it should be for the length of the file (especially thinking ahead if we add other geometry types).

I think it would be easier to follow if each geometry had it's own file, partially because you could more easily compare between geometries, and it would separate geometry specific logic from common shared logic.

```
- select_common.py
- area.py
- position.py
- query_params.py
```

**Comment by mpiannucci** [comment]
Agree with this feedback, especially because I want to add multipoint support next which further complicates things

File: xpublish_edr/plugin.py (hunk 5 above)
Comment (abkfenris) [reply xpublish_edr/plugin.py:108]: I'd also be tempted to reshuffle `select_query` to be a method on `EDRQuery`.
```suggestion
                ds = edr_query.select(ds, dict(request.query_params))
```

**Comment by mpiannucci** [comment]
@abkfenris Let me know how you feel about this reorg
